cas-server-core/src/main/java/org/jasig/cas/services/DefaultRegisteredServiceAuthorizationStrategy.java

```diff
@@ -0,0 +1,258 @@
+     * </ul>
+     * Note that comparison of principal/required attributes is case-sensitive. Exact matches are required
+     * for any individual attribute value.
+     */
+    @Override
+    public boolean isServiceAccessAuthorizedForPrincipal(final Map<String, Object> principalAttributes, final Service service) {
+        if (this.requiredAttributes.isEmpty()) {
+            logger.debug("No required attributes are specified");
+            return true;
+        }
+        if (principalAttributes.isEmpty()) {
+            logger.warn("No principal attributes are found to satisfy attribute requirements for [{}]", service.getId());
+            return false;
+        }
+
+        if (principalAttributes.size() < this.requiredAttributes.size()) {
+            logger.warn("The size of the principal attributes that are [{}] does not match requirements, "
+                    + "which means the principal is not carrying enough data to grant authorization",
+                    principalAttributes);
+            return false;
+        }
+
+        logger.debug("These required attributes [{}] are examined against [{}] before service [{}] can proceed.",
+                getRequiredAttributes(), principalAttributes, service.getId());
+
+        final Sets.SetView<String> difference = Sets.intersection(getRequiredAttributes().keySet(), principalAttributes.keySet());
+        final Set<String> copy = difference.immutableCopy();
+
+        if (this.requireAllAttributes && copy.size() < this.requiredAttributes.size()) {
+            logger.warn("Not all required attributes are available to the principal");
+            return false;
+        }
+
+        for (final String key : copy) {
+            final Set<?> requiredValues = this.requiredAttributes.get(key);
+            Set<?> availableValues;
+
+            final Object objVal = principalAttributes.get(key);
+            if (objVal instanceof Collection) {
+                final Collection valCol = (Collection) objVal;
+                availableValues = Sets.newHashSet(valCol.toArray());
+            } else {
+                availableValues = Collections.singleton(objVal);
+            }
+
+            final Sets.SetView<?> differenceInValues = Sets.intersection(availableValues, requiredValues);
+            if (differenceInValues.size() > 0) {
+                logger.info("Principal is authorized to access service [{}]", service.getId());
+                return true;
+            }
+        }
+        logger.info("Principal is denied access to service [{}]", service.getId());
+        return false;
+    }
+
+    @Override
+    public boolean isServiceAuthorizedForSso(@NotNull final Service service) {
+        if (!this.ssoEnabled) {
+            logger.warn("Service [{}] is not allowed to participate in SSO.", service.getId());
+        }
+        return this.ssoEnabled;
+    }
+
+    @Override
+    public boolean isServiceAuthorized(final Service service) {
```

Comment:
if we're attaching these strategies to a service it seems weird to require said service to be passed in.